### PR TITLE
independent utility to write metadata file after the run completes

### LINKED ISSUE #103: Independent function for writing the nx hdf file after acquisition is complete
We would like to have an independent function for writing nexus file after the acquisition function (containing `bp.count`) is called. We would prefer to have a plain python function independent from Bluesky RE (and `apstools`, if possible) so that it is easier for beamline folks to develop, maintain and debug. 

If we can have only just a template showing how it's done, I can populate the rest of the metadata fields in the hdf file upon request from Suresh and users.

### PR DESCRIPTION
**Pull request by prjemian**
- close #103

### CODE REVIEW

File: src/instrument/utils/conversions.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+    ~ts2iso
+"""
+
+import datetime
+
+
+def ts2iso(ts: float, sep: str = " ") -> str:
+    """Convert Python timestamp (float) to IS8601 time in current time zone."""
+    return datetime.datetime.fromtimestamp(ts).isoformat(sep=sep)
```
Comment (MDecarabas) [inline src/instrument/utils/conversions.py:14]: Any reason why this is now in its own file? Also how often are we using this function that we need it to be it's own thing?

File: src/instrument/utils/dm_metadata_file.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import pathlib
+
+import databroker
+import h5py
+import yaml
+
+from .conversions import ts2iso
+from .iconfig_loader import iconfig  # noqa
+
+CATALOG = iconfig.get("DATABROKER_CATALOG", "8idi_xpcs")
```
Comment (MDecarabas) [inline src/instrument/utils/dm_metadata_file.py:18]: Why is 8idi_xpcs hard coded, isn't it already written in the iconfig?

File: src/instrument/utils/dm_metadata_file.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+XPCS_LAYOUT_VERSION = "APS-8IDI-202310"
```
Comment (MDecarabas) [inline src/instrument/utils/dm_metadata_file.py:19]: What does this showcase?

File: src/instrument/utils/conversions.py (hunk 1 above)
Comment (prjemian) [reply src/instrument/utils/conversions.py:14]: Already existed in another file, which also imported from `bluesky`.  Original statement included:

> We would prefer to have a plain python function independent from Bluesky RE

File: src/instrument/utils/dm_metadata_file.py (hunk 2 above)
Comment (prjemian) [reply src/instrument/utils/dm_metadata_file.py:18]: From testing

File: src/instrument/utils/dm_metadata_file.py (hunk 3 above)
Comment (prjemian) [reply src/instrument/utils/dm_metadata_file.py:19]: `XPCS_LAYOUT_VERSION` is for the consumer of this file, to expect that certain items are present.

File: src/instrument/utils/dm_metadata_file.py (hunk 2 above)
Comment (MDecarabas) [reply src/instrument/utils/dm_metadata_file.py:18]: sounds good. Guess it's good to be removed then?

File: src/instrument/utils/dm_metadata_file.py (hunk 3 above)
Comment (MDecarabas) [reply src/instrument/utils/dm_metadata_file.py:19]: Nice, is there anywhere for users to see what layout versions are expected and the options one could use?

Comment (prjemian) [reply src/instrument/utils/dm_metadata_file.py:19]: No, this is a conversation with @AZjk .

**Comment by qzhang234** [comment]
@prjemian @sureshnaps Here is an example for creating md and converting md to Nexus:
https://github.com/aps-8id-dys/bluesky/blob/d71412db4f628d9aa60ebdba1f9622372fab8c0e/src/instrument/plans/simple_plans.py#L39-L72
Currently Eiger4M is powered off so I put detector as None.

Please see if all of you approve. Upon your go-ahead I will replace this block of code with a single `bp.count[det], md)`, and then I'll work with @AZjk to make sure the Nexus file works with his XPCS code:
https://github.com/aps-8id-dys/bluesky/blob/d71412db4f628d9aa60ebdba1f9622372fab8c0e/src/instrument/plans/simple_plans.py#L130-L135

**Comment by qzhang234** [comment]
Please see if all of you approve. Upon your go-ahead I will replace this block of code with a single `bp.count[det], md)`, and then I'll work with @AZjk to make sure the Nexus file works with his XPCS code:
https://github.com/aps-8id-dys/bluesky/blob/d71412db4f628d9aa60ebdba1f9622372fab8c0e/src/instrument/plans/simple_plans.py#L130-L135

**Comment by prjemian** [comment]
That writes an hdf5 file.  Needs a little more to make it a NeXus file. I'll show you what to add.

**Comment by prjemian** [comment]
@qzhang234  : Consider #108 instead of this PR.  If you prefer to work with #108, then close this PR without merging.  

Continue to add content like these lines: https://github.com/aps-8id-dys/bluesky/blob/edafabe8beb217b0968cf76523fca6541d464e26/src/instrument/plans/simple_plans.py#L71-L72

**Comment by qzhang234** [comment]
Hi @prjemian, could you elaborate what `.attrs["NX_class"]` do in the lines you added? And I assume for every value I add to the Nexus file, I will need to specify the `NX_class`, right? How many `NX_class` are there?

Also, is this the file format that Bluesky nxwriter is currently using to write the files we have collected in the past two weeks?

**Comment by prjemian** [comment]
`NX_class`: https://manual.nexusformat.org/datarules.html#id2

**Comment by prjemian** [comment]
@AZjk : Please confirm for @qzhang234 that all the items you have [listed](https://github.com/aps-8id-dys/bluesky/issues/103#issuecomment-2458049361) belong in the `/entry/instrument/bluesky/metadata/` group of the HDF5 file.

**Comment by qzhang234** [comment]
Hi @prjemian , the links provided explained what a `NX_class` attribute is, but not how it is used and why it is necessary. Could you elaborate on that? What's the disadvantage of using plain hdf files?

**Comment by prjemian** [comment]
As the NeXus manual explains (not repeating that here), these are required by the NeXus standard to define the structure of a NeXus file.  Without them, the file is not recognized by NeXus; it's just another HDF5 file with no standard structure.

Your group agreed to use a NeXus file here.

**Comment by prjemian** [comment]
A little bit more [searching](https://manual.nexusformat.org/search.html?q=NX_class) of the NeXus manual points to [thissection](https://manual.nexusformat.org/examples/python/simple_example_basic/index.html#write-a-nexus-hdf5-file):

> ... our support library must create and set the `NX_class` attribute on each group.

**Comment by prjemian** [comment]
To be superseded by newer work in branch [`103-qz-nexus-metadata-writer`](https://github.com/aps-8id-dys/bluesky/tree/103-qz-nexus-metadata-writer).

**Comment by qzhang234** [comment]
Hi @prjemian, can this pull request also be closed and the branch deleted?

**Pull request closed without merging**